lucene/core/src/java/org/apache/lucene/util/AttributeSource.java

```diff
@@ -171,6 +171,23 @@ public boolean hasNext() {
     return implInterfaces.get(clazz);
   }
   
+  /**
+   * Returns the instance of the passed in AttributeImpl contained in this AttributeSource
+   * <p>
+   * The caller must pass in a Class&lt;? extends AttributeImpl&gt; value. 
+   *
+   * @return instance of the passed in AttributeImpl, or {@code null} if this AttributeSource 
+   *         does not contain the AttributeImpl. It is recommended to always use 
+   *         {@link #addAttributeImpl} even in consumers  of TokenStreams, because you cannot 
+   *         know if a specific TokenStream really uses a specific AttributeImpl. 
+   *         {@link #addAttributeImpl} will automatically make the attribute impl available. 
+   *         If you want to only use the attribute , if it is available (to optimize
+   *         consuming), use {@link #hasAttribute}.
+   */
+  public final <T extends AttributeImpl> T getAttributeImpl(Class<T> attClass) {
+    return attClass.cast(attributeImpls.get(attClass));
+  }
+  
```

Comment:
This shouldn't be necessary, we use interfaces everywhere else?

Reply:
I am pretty sure , when I was debugging that at the moment that is the only way to get the Payload attributes (they seem to be there just in the attr impl list.
But I can double check this!